src/domain/auth-rebac/domain/src/repos/rebac_repository.rs

```diff
@@ -0,0 +1,241 @@
+        subject_entity: &Entity,
+    ) -> Result<Vec<ObjectEntityWithRelation>, GetEntityRelationsError>;
+
+    async fn get_subject_entity_relations_by_object_type(
+        &self,
+        subject_entity: &Entity,
+        object_entity_type: EntityType,
+    ) -> Result<Vec<ObjectEntityWithRelation>, GetEntityRelationsError>;
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetEntityRelationsError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Errors
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum UpsertEntityPropertyError {
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum DeleteEntityPropertyError {
+    #[error(transparent)]
+    NotFound(EntityPropertyNotFoundError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+#[error("Entity property not found: {self}")]
+pub struct EntityPropertyNotFoundError {
+    pub entity_type: String,
+    pub entity_id: String,
+    pub property_name: String,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum GetEntityPropertiesError {
+    #[error(transparent)]
+    NotFound(EntityPropertiesNotFoundError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+#[error("Entity properties not found: {self}")]
+pub struct EntityPropertiesNotFoundError {
+    pub entity_type: String,
+    pub entity_id: String,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum InsertEntitiesRelationError {
+    #[error(transparent)]
+    Duplicate(InsertEntitiesRelationDuplicateError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+#[error("Duplicate entity relation not inserted: {self}")]
+pub struct InsertEntitiesRelationDuplicateError {
+    pub subject_entity_type: String,
+    pub subject_entity_id: String,
+    pub relationship: String,
+    pub object_entity_type: String,
+    pub object_entity_id: String,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum DeleteEntitiesRelationError {
+    #[error(transparent)]
+    NotFound(EntitiesRelationNotFoundError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+#[error("Entities relation not found: {self}")]
```

Comment:
Hm, how does this work? I thought `#[error("Blah")]` implements the `Display` trait. By having `{self}` in the error message are you recursively calling display on the error. Will this crash with stack overflow?

Reply:
Oh, you're right...

> Will this crash with stack overflow?

Unequivocally

Reply:
Fixed